app/models/notification.rb

```diff
@@ -3,93 +3,56 @@
+  has_many :actors, class_name: "Person", through: :notification_actors, source: :person
+  belongs_to :target, polymorphic: true
 
   attr_accessor :note_html
 
   def self.for(recipient, opts={})
-    self.where(opts.merge!(:recipient_id => recipient.id)).order('updated_at desc')
-  end
-
-  def self.notify(recipient, target, actor)
-    return nil unless target.respond_to?(:notification_type) && recipient.person != actor
-
-    note_type = target.notification_type(recipient, actor)
-    return nil unless note_type
-
-    return_note = if [Comment, Like, Reshare].any? { |klass| target.is_a?(klass) }
-      s_target = target.is_a?(Reshare) ? target.root : target.parent
-      note_type.concatenate_or_create(recipient, s_target,
-                                          actor, note_type)
-    else
-      note_type.make_notification(recipient, target,
-                                      actor, note_type)
-    end
-    return_note.email_the_user(target, actor) if return_note
-    return_note
+    where(opts.merge!(recipient_id: recipient.id)).order("updated_at desc")
   end
 
   def as_json(opts={})
-    super(opts.merge(:methods => :note_html))
+    super(opts.merge(methods: :note_html))
   end
 
   def email_the_user(target, actor)
-    self.recipient.mail(self.mail_job, self.recipient_id, actor.id, target.id)
+    recipient.mail(mail_job, recipient_id, actor.id, target.id)
   end
 
   def set_read_state( read_state )
     update_column(:unread, !read_state)
   end
 
   def mail_job
-    raise NotImplementedError.new('Subclass this.')
+    raise NotImplementedError.new("Subclass this.")
   end
 
-  def effective_target
-    self.popup_translation_key == "notifications.mentioned" ? self.target.post : self.target
+  def linked_object
+    target
   end
 
-private
-  def self.concatenate_or_create(recipient, target, actor, notification_type)
+  def self.concatenate_or_create(recipient, target, actor)
     return nil if suppress_notification?(recipient, target)
 
-    if n = notification_type.where(:target_id => target.id,
-                                   :target_type => target.class.base_class,
-                                   :recipient_id => recipient.id,
-                                   :unread => true).first
-
-      begin
-        n.actors = n.actors | [actor]
-        n.unread = true
-        # Explicitly touch the notification to update updated_at whenever new actor is inserted in notification.
-        n.touch
-        n.save!
-      rescue ActiveRecord::RecordNotUnique
-        nil
-      end
-      n
-    else
-      make_notification(recipient, target, actor, notification_type)
+    find_or_initialize_by(recipient: recipient, target: target, unread: true).tap do |notification|
+      notification.actors |= [actor]
+      # Explicitly touch the notification to update updated_at whenever new actor is inserted in notification.
+      notification.touch unless notification.new_record?
```

Comment:
Micro: We shouldn't need to `save!` if the record has no changes, just `touch` and we shouldn't need to `touch` if the record has changes, `save!` will do it:
```rb
if notification.new_record? || notification.changed?
  notification.save!
else
  notification.touch
end